babe: treat epoch_authorship RPC method as unsafe

andresilva opened the pull request:
BABE's RPC API exposes a method to calculate all the slots that are allocated to the local keys in the current epoch. This information is sensitive and the method should not be exposed if the RPC is publicly addressable (like we do for authorship APIs).

Most of the changes in this PR are to allow the service builder to pass on a `DenyUnsafe` instance to the RPC extensions builder. Unfortunately the `DenyUnsafe` cannot be created at service instantiation time (since only when we try to actually bind http or ws interfaces do we know if it's unsafe), therefore we need to return a closure that builds the rpc extension. The resulting API (i.e. the new signature of `ServiceBuilder::with_rpc_extensions`) is a bit messy and complicated but doing this with a better API requires a refactor of the `ServiceBuilder` that I can't take on now.

polkadot companion: https://github.com/paritytech/polkadot/pull/1108

DemiMarie-parity commented on client/rpc-api/src/policy.rs:
```diff
@@ -25,37 +25,37 @@ use jsonrpc_core as rpc;
-    No
+	/// Denies only potentially unsafe RPCs.
+	Yes,
+	/// Allows calling every RPCs.
```
```suggestion
	/// Allows calling all RPCs.
```

DemiMarie-parity commented on client/rpc-api/src/policy.rs:
```diff
@@ -25,37 +25,37 @@ use jsonrpc_core as rpc;
-            DenyUnsafe::No => Ok(())
-        }
-    }
+	/// Returns `Ok(())` if the RPCs considered unsafe are safe to call,
```
Should the whitespace changes be a separate PR?

andresilva commented:
The only unrelated whitespace change is in the `policy.rs` file, where the only changes I've made were reformatting with tabs instead of spaces (no content was changed).

tomusdrw commented on client/service/src/builder.rs:
```diff
@@ -679,21 +679,27 @@ impl<TBl, TRtApi, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, TRpc, Backend>
+	where
+		TSc: Clone,
+		TFchr: Clone,
+		URpcBuilder: Fn(sc_rpc::DenyUnsafe) -> URpc + Send + 'static,
```
We could consider traitifying `Fn(sc_rpc::DenyUnsafe) -> URpc + Send + 'static` and implement that trait for both closures and concrete `URpc` type, so that we could maintain backward compatibility. Not sure if it's worth it though.

andresilva replied:
This might be a good idea actually, for example in the light client currently we don't care about `DenyUnsafe` for any rpc extensions. I guess right now anyone who isn't using BABE's RPC wouldn't care that much about it either.

DemiMarie-parity replied:
@tomusdrw how does one do that?  Every time I have tried to do that it fails with some obscure lifetime error.

tomusdrw replied:
You mean this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=59c9d81d289603e857e5a9dbb8970486 ?

tomusdrw commented on bin/node/cli/src/service.rs:
```diff
@@ -99,30 +99,46 @@ macro_rules! new_full_start {
+					.expect("SelectChain is present for full services or set up failed; qed.");
+				let keystore = builder.keystore().clone();
+
+				Ok(move |deny_unsafe| -> RpcExtension {
```
Another way to avoid returning a closure would be to have `deny_unsafe` be a dynamic object that is lazy initialized when RPC is bound. That might require a `Mutex` and `thread_local` cache or some `unsafe` black magic to avoid huge performance costs on every RPC call though.

andresilva replied:
But when we're building the service we don't know how many times the given extension will be initialized (http, ws?), and we'd need separate instances of this dynamic object for each instantiation I think.

tomusdrw replied:
Yeah, that's correct - we do need different parameter depending on the transport the extension will be used for, so this solution is a no-go.

tomusdrw commented on bin/node/cli/src/service.rs:
```diff
@@ -99,30 +99,46 @@ macro_rules! new_full_start {
-			.with_rpc_extensions(|builder| -> std::result::Result<RpcExtension, _> {
-				let babe_link = import_setup.as_ref().map(|s| &s.2)
-					.expect("BabeLink is present for full services or set up failed; qed.");
+			.with_rpc_extensions(|builder| {
```
We could expect a closure that get's a `RpcDependencies` object here instead of `Builder`, which would contain all required stuff and `deny_unsafe`. That would simplify the code a bit (no extra closure), but would make this method special in a way that it's not invoked right away like other `with_*` methods.

andresilva commented:
@tomusdrw I think the main problem is that unlike now, we might need to instantiate the rpc extensions with different parameters (currently we just clone the given extension when binding it to a server). It might be the case that we expose HTTP unsafely but not WS, in which case the RPC extension instances must have different `DenyUnsafe` values.

We also can't save a simple closure that takes `|builder, deny_unsafe|`, since the types of `builder` will change as we call the other extension methods. The only way to do this would be to somehow make `with_rpc_extensions` work differently as you suggested so that it is a terminal method, i.e. it builds the service.

DemiMarie-parity commented:
@andresilva I will also add that IPC transports are virtually always safe to expose all RPCs on.

andresilva commented:
@tomusdrw I pushed a commit to traitify the RPC extension builder. It is still a breaking change as it requires wrapping the previous builder in some default builder implementation, the case were we take `DenyUnsafe` stays unchaged since we can convert from the closure to `RpcExtensionBuilder`. I have not added docs or updated the polkadot PR yet, let me know if you think this approach is good and we should move forward with it (or any other suggestions you might have). I am not sure whether this is worth it or not.

andresilva commented:
I went with the suggestion of creating an `RpcExtensionBuilder` trait and in addition only use this trait in a new method `with_rpc_extensions_builder`, this makes this PR a non-breaking change as only RPC extensions that care about using `DenyUnsafe` need to use the builder. Updated the companion as well and added missing docs.
